feat!: improves Credit Pool to protect from cumulative increasing amount of gaps

knst opened the pull request:
## Issue being fixed or feature implemented
Each Asset Unlock transaction requires an unique `index`. This index can't be re-used so far as I know because no quorum can re-sign message with same `requestID`, but 2 txes with same indexes will have same `requestID`.

Normally, most of asset-unlock txes should be mined, but some of them won't be mined, for example, by various reasons:`fee` is too low, trying to unlock more that current limits for block, or any other reason. Asset unlock transactions also permanently expires by design when height of blockchain achieves some numbers since this unlock-tx is signed. But the specific index is already occupied.
It means, that amount of gaps in a datastructure that keep all indexes will be increased over time continuously and earlier or later limit in 10k gaps in indexes could be achieved. Maybe it will take 10 years to achieve this day, may be it will take just a day due to some bug in platform.
Let's take a precaution and update storage by this PR.

Instead of old way with limited storage capacity (no more than 10k for total size of all gaps), let's keep indexes as an ordered list of ranges. For example, the set of these indexes: {1, 2, 3, 5, 7, 8, 9, 1000, 1001} would be presented like this in memory and in evo db:
`{[1, 3], [5, 5], [7, 9], [1000, 1001]}.`

For reference, CSkipList keep this list like: `{all indexes below 1001 is mined except: 4, 6, 10, 11, 12, 13, 14, ... , 999}`

NOTE: this changes may become useless if platform will generate Asset Unlock transaction by other way when `indexes` of expired transactions can be re-used. For now it is impossible so far as I know.

## What was done?
Implemented a new datastructure `CRangesSet` to replace `CSkipSet` and used in CreditPool's impementation in core.
Also updated related unit test.

## How Has This Been Tested?
1. Run the old version of `feature_asset_locks.py` -> it failed as expected because transaction that should generate too big gap successfully mined. 
2. Updated functional test accordingly new behavior -> it succeed now.
3. Added new Unit-Test that is copied from `CSkipList`
 
## Breaking Changes
It changes behavior of Credit Pool for case of big gaps. Won't affect anything at the moment, because it is theoretical situation in future.


## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

UdjinM6 commented on src/test/util_tests.cpp:
```diff
@@ -2163,6 +2164,34 @@ BOOST_AUTO_TEST_CASE(test_Capitalize)
+    std::mt19937 gen;
+    for (size_t test = 0; test < 17; ++test) {
+        std::uniform_int_distribution<uint64_t> dist_value(0, (1 << test));
+        size_t skip_size = test ? (1 << (test - 1)) : 1;
```
ci complains:
```
est/util_tests.cpp: In member function ‘void util_tests::test_CRanges::test_method()’:
test/util_tests.cpp:2172:16: error: unused variable ‘skip_size’ [-Werror=unused-variable]
 2172 |         size_t skip_size = test ? (1 << (test - 1)) : 1;
      |                ^~~~~~~~~
```

UdjinM6 commented:
also, pls rebase on top of the most recent `develop` to fix `ERROR: Uploading artifacts as "archive" to coordinator... 413 Payload Too Large  id=4784634216 responseStatus=413 Payload Too Large status=413 token=64_Cify2` in tsan test job

UdjinM6 commented:
pls see 1c9e21e783

knst commented:
> pls see https://github.com/dashpay/dash/commit/1c9e21e783f267b559f0c63c6964dca1ea89c8d0

I updated based on your diff and added couple more changes:
 - shortened Remove() method similarly to Add()'s improvements.
 - comments are edited
 - re-written `operator<()` (same logic)

UdjinM6 commented on src/evo/creditpool.h:
```diff
@@ -34,7 +34,7 @@ struct CCreditPool {
     CAmount currentLimit{0};
     CAmount latelyUnlocked{0};
-    CSkipSet indexes{};
+    CRangesSet indexes{};
```
Note: CSkipSet is not used anywhere now, could probably drop it

knst replied:
> Note: CSkipSet is not used anywhere now, could probably drop it
For sure, just thought to do it in a spare PR

PastaPastaPasta replied:
dead code should be dropped

knst replied:
dead-code is removed, please, review @PastaPastaPasta

UdjinM6 reviewed (approved):
LGTM, utACK.

knst commented:
resolved conflict due to https://github.com/dashpay/dash/pull/5526 by rebasing to top of develop

UdjinM6 reviewed (dismissed):
re-utACK

PastaPastaPasta commented:
Couldn't we just also remove the limit or something like that ish? I don't understand why ranges are better than "skipped"?

PastaPastaPasta reviewed (changes requested):
Asked questions

knst commented:
> Couldn't we just also remove the limit or something like that ish? I don't understand why ranges are better than "skipped"?

We can't just remove limit. If by any reason the platform will issue Asset Unlock transaction with index 1e9, all dash clients will immediately allocate 8Gb RAM to enumerate billion of meaningless continuous numbers.

It looks like Denial-of-Service for me and most nodes will die after that. That's why limit of capacity has been introduced at first.

UdjinM6 reviewed (approved):
re-utACK

PastaPastaPasta reviewed (approved):
utACK for squash merge
